net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +65,617 @@ impl UdpSocket {
         Self::bind_raw(addr)
     }
 
+    /// Is the socket broken and needs a rebind?
+    pub fn is_broken(&self) -> bool {
+        self.is_broken.load(std::sync::atomic::Ordering::SeqCst)
+    }
+
+    /// Marks this socket as needing a rebind
+    pub fn mark_broken(&self) {
+        self.is_broken
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    /// Rebind the underlying socket.
+    pub fn rebind(&self) -> Result<()> {
+        debug!("rebinding {}", self.addr);
+        // Remove old socket
+        let mut guard = self.socket.write().unwrap();
+        {
+            let socket = guard.take().expect("not yet dropped");
+            drop(socket);
+        }
+
+        // Prepare new socket
+        let new_socket = inner_bind(self.addr)?;
```

Comment:
note that if rebind fails, the inner socket will be `None` when not dropped and a bunch of `expect`s will scream

Reply:
fixed